src/main/java/io/confluent/kafkarest/ConsumerReadTask.java

```diff
@@ -108,37 +110,50 @@ public boolean doPartialRead() {
       long startedIteration = parent.getConfig().getTime().milliseconds();
       final int requestTimeoutMs = parent.getConfig().getInt(
           KafkaRestConfig.CONSUMER_REQUEST_TIMEOUT_MS_CONFIG);
+      final long endTime = startedIteration + requestTimeoutMs;
+      final int itBackoff = parent.getConfig().getInt(
+        KafkaRestConfig.CONSUMER_ITERATOR_BACKOFF_MS_CONFIG);
       try {
-        // Read off as many messages as we can without triggering a timeout exception. The
-        // consumer timeout should be set very small, so the expectation is that even in the
-        // worst case, num_messages * consumer_timeout << request_timeout, so it's safe to only
-        // check the elapsed time once this loop finishes.
-        while (iter.hasNext()) {
-          MessageAndMetadata<KafkaK, KafkaV> msg = iter.peek();
-          ConsumerRecordAndSize<ClientK, ClientV> recordAndSize = parent.createConsumerRecord(msg);
-          long roughMsgSize = recordAndSize.getSize();
-          if (bytesConsumed + roughMsgSize > maxResponseBytes) {
-            break;
+        while (parent.getConfig().getTime().milliseconds() < endTime) {
+
+          if (iter == null || !iter.hasNext()) {
+            // The consumer timeout should be set very small, so the expectation is that even in the
+            // worst case, num_messages * consumer_timeout << request_timeout, so it's safe to only
+            // check the elapsed time once this loop finishes.
+            ConsumerRecords<KafkaK, KafkaV> records = consumer.poll(itBackoff);
```

Comment:
The 0.8 Kafka Consumer has two settings,  one is the num.consumer.fetchers and the other is queued.max.message.chunks, configuring these settings approach greatly improves the performance in the REST approach as once a request has been served for a consumer, it starts fetching messages in the background, it already has some messages fetched by the time another request comes in for the consumer instance. With 0.9 Kafka Consumer, we are kinda of losing this which will definitely reduce the performance/throughput. I would suggest to create a wrapper around the KafkaConsumer, that maintains a blocking queue (configurable) and the keeps polling for new messages in the background rather than waiting for a request to arrive, which also solves the problem with heartbeats/rebalancing that @hachikuji was talking about.  This wrapper maintains internal state of offsets to be committed, similar to the 0.8 Kafka Consumer